addons/misra.py

```diff
@@ -1494,6 +1494,254 @@ def misra_8_14(self, rawTokens):
             if token.str == 'restrict':
                 self.reportError(token, 8, 14)
 
+    def misra_9_2(self, data):
+        # Holds information about a struct or union's element definition.
+        class ElementDef:
+            def __init__(self, elementType, name, valueType = None, dimensions = None):
+                self.elementType = elementType
+                self.name = name
+                self.valueType = valueType
+                self.dimensions = dimensions
+
+        def getIntvalue(token):
+            if token and token.values and len(token.values) > 0:
+                return token.values[0].intvalue
+            else:
+                return None
+
+        # Return an array containing the size of each dimension of an array declaration,
+        # or coordinates of a designator in an array initializer,
+        # and the name token's valueType, if it exist.
+        #
+        # In the examples below, the ^ indicates the initial token passed to the function.
+        #
+        # Ex:   int arr[1][2][3] = .....
+        #                    ^
+        #       returns: [1,2,3], valueType
+        #
+        # Ex:   int arr[3][4] = { [1][2] = 5 }
+        #                            ^
+        #       returns [1,2], None
+        def getArrayDimensionsAndValueType(token):
+            dimensions = []
+            while token and token.str == '[':
+                op1Value = getIntvalue(token.astOperand1)
+                op2Value = getIntvalue(token.astOperand2)
+                if op2Value != None:
+                    dimensions.insert(0, op2Value)
+                    token = token.astOperand1
+                elif op1Value != None:
+                    dimensions.insert(0, op1Value)
+                    break
+                else:
+                    dimensions = None
+                    break
+
+            if token.valueType:
+                valueType = token.valueType
+            else:
+                valueType = None
+
+            return dimensions, valueType
+
+        # Returns a list of the struct elements as StructElementDef in the order they are declared.
+        def getRecordElements(valueType):
+            elements = []
+            for token in data.tokenlist:
```

Comment:
hmm.. this is not how you should do it. This would have been better:

    for var in valueType.typeScope.varlist:
        token = var.nameToken
        ....

But I am a bit confused .. for some reason the `Scope` does not seem to have a `varlist`. However the dump file has the `varlist` info needed. So somebody should add a `varlist` in the Scope class.

Reply:
For this testcode:
```
struct A {
    int x;
    int y;
};
```
The dump file will contain:
```
    <scope id="0x560e78458260" type="Struct" className="A" bodyStart="0x560e784532b0" bodyEnd="0x560e78456020" nestedIn="0x560e78458040">
      <varlist>
        <var id="0x560e784567a0"/>
        <var id="0x560e78457c10"/>
      </varlist>
    </scope>
```

Reply:
We saw the varlist in the dump file, but also realized that it wasn't present in the Scope objects. So we then tried looping though the variables list, but found they weren't always in order, so we resigned to using the tokenList for now.

Reply:
ok.

I have the feeling that variables in `<variables>`  might not be ordered always because the variables are created in a few steps.

However I am spontaneously surprised if `<varlist>` is not ordered.

To be clear.. you saw that `<varlist>` was not ordered? do you have an example code?

Reply:
I have added `Scope.varlist`. Can you please try if this works:

    for var in valueType.typeScope.varlist:
        token = var.nameToken
        ....

Reply:
Nice, this feels better. I've changed our code to use this new list.

Btw: It was the data.variables list that was unordered. Scope.varlist has behaved very orderly thus far